packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp

```diff
@@ -506,175 +508,229 @@ void CoalesceDropFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(Level
+			auto rowView = A_device.row(row);
+			size_t nnz = rowView.length;
+
+			size_t n = 0;
+			auto drop_view = Kokkos::subview(drop_views, Kokkos::make_pair(A_device.graph.row_map(row), A_device.graph.row_map(row+1)));
+			auto index_view = Kokkos::subview(index_views, Kokkos::make_pair(A_device.graph.row_map(row), A_device.graph.row_map(row+1)));
+
+			//find magnitudes
+			Kokkos::parallel_reduce(Kokkos::TeamThreadRange(teamMember, (LO)nnz), [&](const LO colID, size_t &count) {
+				index_view(colID) = colID;
+				LO col = rowView.colidx(colID);
+				if(row == col) {
+					drop_view(colID) = true;
+					count++;
+				}
+				//Don't aggregate boundaries
+				else if(boundaryNodesDevice(colID)) {
+					drop_view(colID) = true;
+				}
+				else {
+					drop_view(colID) = false;
+					count++;
+				}
+			}, n);
+
+			size_t dropStart = n;
+			if (algorithm == unscaled_cut) {
+				Kokkos::Experimental::sort_team(teamMember, index_view, [=](size_t& x, size_t& y) {
+					if(drop_view(x) || drop_view(y)) {
+						return drop_view(x) < drop_view(y);
+					}
+					else {
+						typename STS::magnitudeType x_aij    = static_cast<SC>(std::fabs(static_cast<double>(rowView.value(x) * rowView.value(x))));
+						typename STS::magnitudeType y_aij    = static_cast<SC>(std::fabs(static_cast<double>(rowView.value(y) * rowView.value(y))));
+						return x_aij > y_aij;
+					}
+				});
+
+				//find index where dropping starts
+				Kokkos::parallel_reduce(Kokkos::TeamThreadRange(teamMember, 1, n), [=](size_t i, size_t& min) {
+					auto const& x = index_view(i - 1);
+					auto const& y = index_view(i);
+					typename STS::magnitudeType x_aij = 0;
+					typename STS::magnitudeType y_aij = 0;
+					if(!drop_view(x)) {
+						x_aij    = static_cast<SC>(std::fabs(static_cast<double>(rowView.value(x) * rowView.value(x))));
+					}
+					if(!drop_view(y)) {
+						y_aij    = static_cast<SC>(std::fabs(static_cast<double>(rowView.value(y) * rowView.value(y))));
+					}
+
+					if(x_aij > realThreshold * y_aij) {
+						if(i < min) {
+							min = i;
+						}
+					}
+				}, Kokkos::Min<size_t>(dropStart));
+          	 	} else if (algorithm == scaled_cut) {
+				Kokkos::Experimental::sort_team(teamMember, index_view, [=](size_t& x, size_t& y) {
+					if(drop_view(x) || drop_view(y)) {
+						return drop_view(x) < drop_view(y);
+					}
+					else {
+						typename STS::magnitudeType x_aij    = static_cast<SC>(std::fabs(static_cast<double>(rowView.value(x) * rowView.value(x))));
+						typename STS::magnitudeType y_aij    = static_cast<SC>(std::fabs(static_cast<double>(rowView.value(y) * rowView.value(y))));
+						typename STS::magnitudeType x_aiiajj = static_cast<SC>(std::fabs(static_cast<double>(threshold * threshold * ghostedDiagValsView(rowView.colidx(x)) * ghostedDiagValsView(row))));
+						typename STS::magnitudeType y_aiiajj = static_cast<SC>(std::fabs(static_cast<double>(threshold * threshold * ghostedDiagValsView(rowView.colidx(y)) * ghostedDiagValsView(row))));
```

Comment:
`std::fabs` might need to be replaced with `Kokkos::ArithTraits<impl_scalar_type>::magnitude`. We need to think through this a bit for complex scalar type. Same for the other spots where we cast things around.